### PR TITLE
[781] Add support for multiline labels

### PR DESCRIPTION
**Pull request by lfasani**
### Type of this PR 

- [ ] Bug fix
- [ X ] New feature or improvement
- [ ] Documentation
- [ ] Cleanup
- [ ] Test
- [ ] Build/Releng

### Issue(s)

...

### What does this PR do?

...

### Screenshot/screencast of this PR

...
 

### Potential side effects

...

### How to test this PR?

- [ ] Frontend Unit tests
- [ ] Backend Unit tests
- [ ] Cypress : please specify
- [ ] Manual Test : please specify

### Checklist

- [ ] I have read CONTRIBUTING carefully.
- [ ] I have signed the [Contributor License Agreement](http://www.eclipse.org/legal/CLA.php).
- [ ] All my commits are signed-off (`-s`) with my mail address of my Eclipse Account.
- [ ] I have covered my changes by unit tests or integration tests or manual tests.
- [ ] All tests pass.
- [ ] I have updated the documentation accordingly
- [ ] New React components are availables in storybook

### CODE REVIEW

File: backend/sirius-components-diagrams-layout/src/main/java/org/eclipse/sirius/components/diagrams/layout/ELKLayoutedDiagramProvider.java [hunk 1]
```diff
@@ -76,11 +76,17 @@ private Node getLayoutedNode(Node node, ElkConnectableShape elkConnectableShape,
...
         Position position = Position.at(elkConnectableShape.getX(), elkConnectableShape.getY());
 
-        // @formatter:off
-        Label label = this.getLayoutedLabel(node.getLabel(), id2ElkGraphElements, 0, 0);
+        double xOffSet = 0;
+        if (!node.isBorderNode()) {
+            // The label is positioned at the center of the node and the front-end will apply a "'text-anchor':
+            // 'middle'" property.
+            xOffSet = node.getLabel().getSize().getWidth() / 2;
+        }
```
Comment (sbegaudeau) [inline backend/sirius-components-diagrams-layout/src/main/java/org/eclipse/sirius/components/diagrams/layout/ELKLayoutedDiagramProvider.java:84]: Can make this hypothesis if in the class below, we have the code to deal with various label placement strategies? What was the issue with the previous algorithm?

File: backend/sirius-components-diagrams-layout/src/test/java/org/eclipse/sirius/components/diagrams/layout/incremental/DiagramLayoutTest.java [hunk 2]
```diff
@@ -159,4 +160,40 @@ public void testSimpleDiagramLayout() throws IOException {
...
+                .rectangleNode(nodeLabelWithMultiple).at(10, 10).of(10, 10)
+                    .childNodes()
+                        .rectangleNode(firstChildTargetObjectId).at(10, 10).of(50, 50).and()
+                    .and()
+                .and()
+            .and()
+        .build();
+        // @formatter:on
+
+        Path path = Paths.get("src", "test", "resources", "editing-contexts", "testSimpleDiagramLayout"); //$NON-NLS-1$//$NON-NLS-2$//$NON-NLS-3$//$NON-NLS-4$ //$NON-NLS-5$
```
Comment (sbegaudeau) [inline backend/sirius-components-diagrams-layout/src/test/java/org/eclipse/sirius/components/diagrams/layout/incremental/DiagramLayoutTest.java:181]: Don't reuse test data from other tests, create your own

File: backend/sirius-components-diagrams-layout/src/test/java/org/eclipse/sirius/components/diagrams/layout/incremental/DiagramLayoutTest.java [hunk 3]
```diff
@@ -90,8 +91,8 @@ private TestDiagramCreationService createDiagramCreationService(Diagram diagram)
         NodeSizeProvider nodeSizeProvider = new NodeSizeProvider(new ImageSizeProvider());
         IncrementalLayoutEngine incrementalLayoutEngine = new IncrementalLayoutEngine(nodeSizeProvider);
 
-        LayoutService layoutService = new LayoutService(new IELKDiagramConverter.NoOp(), new IncrementalLayoutDiagramConverter(), new LayoutConfiguratorRegistry(List.of()),
-                new ELKLayoutedDiagramProvider(), new IncrementalLayoutedDiagramProvider(), representationDescriptionSearchService, incrementalLayoutEngine);
+        LayoutService layoutService = new LayoutService(new ELKDiagramConverter(new TextBoundsService(), new ImageSizeProvider()), new IncrementalLayoutDiagramConverter(),
+                new LayoutConfiguratorRegistry(List.of()), new ELKLayoutedDiagramProvider(), new IncrementalLayoutedDiagramProvider(), representationDescriptionSearchService, incrementalLayoutEngine);
```
Comment (sbegaudeau) [inline backend/sirius-components-diagrams-layout/src/test/java/org/eclipse/sirius/components/diagrams/layout/incremental/DiagramLayoutTest.java:95]: Why is that necessary? Do you really need an ImageSizeProvider?

File: backend/sirius-components-diagrams-layout/src/test/java/org/eclipse/sirius/components/diagrams/layout/incremental/DiagramLayoutTest.java [hunk 4]
```diff
@@ -159,4 +160,40 @@ public void testSimpleDiagramLayout() throws IOException {
...
+        .build();
+        // @formatter:on
+
+        Path path = Paths.get("src", "test", "resources", "editing-contexts", "testSimpleDiagramLayout"); //$NON-NLS-1$//$NON-NLS-2$//$NON-NLS-3$//$NON-NLS-4$ //$NON-NLS-5$
+        JsonBasedEditingContext editingContext = new JsonBasedEditingContext(path);
+
+        TestDiagramCreationService diagramCreationService = this.createDiagramCreationService(diagram);
+
+        Diagram layoutedDiagram = diagramCreationService.performElKLayout(editingContext, diagram);
+
```
Comment (sbegaudeau) [inline backend/sirius-components-diagrams-layout/src/test/java/org/eclipse/sirius/components/diagrams/layout/incremental/DiagramLayoutTest.java:187]: Why would we test the ELK layout in the `incremental` layout tests?

File: backend/sirius-components-diagrams-layout/src/test/java/org/eclipse/sirius/components/diagrams/layout/services/TestDiagramCreationService.java [hunk 5]
```diff
@@ -94,4 +94,8 @@ public Diagram performLayout(IEditingContext editingContext, Diagram diagram, ID
         return this.layoutService.incrementalLayout(editingContext, diagram, Optional.of(diagramEvent));
     }
 
+    public Diagram performElKLayout(IEditingContext editingContext, Diagram diagram) {
+        return this.layoutService.layout(editingContext, diagram);
+    }
+
```
Comment (sbegaudeau) [inline backend/sirius-components-diagrams-layout/src/test/java/org/eclipse/sirius/components/diagrams/layout/services/TestDiagramCreationService.java:100]: Do we really need that?

File: backend/sirius-components-diagrams/src/main/java/org/eclipse/sirius/components/diagrams/TextBoundsProvider.java [hunk 6]
```diff
@@ -58,10 +59,23 @@ public TextBounds computeBounds(LabelStyle labelStyle, String text) {
...
+            for (int i = 1; i < lines.length; i++) {
+                String line = lines[i];
+
+                Rectangle2D lineBounds = font.getStringBounds(line, FONT_RENDER_CONTEXT);
+                // shift the rectangle under the previous line
+                lineBounds.setFrame(lineBounds.getX(), lineBounds.getY() + labelBounds.getHeight(), lineBounds.getWidth(), lineBounds.getHeight());
+
+                labelBounds = labelBounds.createUnion(lineBounds);
+            }
+        }
```
Comment (sbegaudeau) [inline backend/sirius-components-diagrams/src/main/java/org/eclipse/sirius/components/diagrams/TextBoundsProvider.java:75]: what if `lines.length == 0`?

File: backend/sirius-components-diagrams/src/main/java/org/eclipse/sirius/components/diagrams/components/LabelType.java [hunk 7]
```diff
@@ -1,5 +1,5 @@
 /*******************************************************************************
- * Copyright (c) 2021 Obeo.
+ * Copyright (c) 2021,2022 Obeo.
```
Comment (sbegaudeau) [inline backend/sirius-components-diagrams/src/main/java/org/eclipse/sirius/components/diagrams/components/LabelType.java:2]: missing space

File: backend/sirius-components-emf/src/main/java/org/eclipse/sirius/components/emf/compatibility/properties/EStringIfDescriptionProvider.java [hunk 8]
```diff
@@ -39,7 +39,7 @@
 public class EStringIfDescriptionProvider {
     private static final String IF_DESCRIPTION_ID = "EString"; //$NON-NLS-1$
 
-    private static final String TEXTFIELD_DESCRIPTION_ID = "Textfield"; //$NON-NLS-1$
+    private static final String TEXTAREA_DESCRIPTION_ID = "Textarea"; //$NON-NLS-1$
```
Comment (sbegaudeau) [inline backend/sirius-components-emf/src/main/java/org/eclipse/sirius/components/emf/compatibility/properties/EStringIfDescriptionProvider.java:42]: Why does this file needs to be changed? Using a textfield does not change anything at all on the behavior of the frontend nor on this feature, right?

File: frontend/src/diagram/sprotty/DependencyInjection.ts [hunk 9]
```diff
@@ -66,27 +68,66 @@ import {
...
-export class SEditableLabel extends SLabel {
-  initialText: string;
-  preSelect: boolean = true;
-}
-
 class EditLabelUIWithInitialContent extends EditLabelUI {
   protected applyTextContents() {
-    if (this.label instanceof SEditableLabel) {
-      this.inputElement.value = this.label.initialText || this.label.text;
+    if (this.label instanceof Label) {
```
Comment (sbegaudeau) [inline frontend/src/diagram/sprotty/DependencyInjection.ts:73]: This class is starting to get big, could you extract it in a dedicated file please.

File: frontend/src/diagram/sprotty/views/LabelView.tsx [hunk 10]
```diff
@@ -48,12 +49,15 @@ export class LabelView extends SLabelView {
         styleObject['text-decoration'] += ' line-through';
       }
     }
+    if (label.type?.includes('center')) {
```
Comment (sbegaudeau) [inline frontend/src/diagram/sprotty/views/LabelView.tsx:52]: Can the type really be null? Does that mean that the backend does not have the precise position of the label anymore? For example to compute overlap and things like that.

File: frontend/src/diagram/sprotty/views/LabelView.tsx [hunk 11]
```diff
@@ -64,4 +68,22 @@ export class LabelView extends SLabelView {
     }
     return vnode;
   }
+  convertText(text: string, fontSize: number) {
```
Comment (sbegaudeau) [inline frontend/src/diagram/sprotty/views/LabelView.tsx:71]: What is the return type of this function? Is that some JSX elements? If that's the case, then you have a function which takes some arguments or props and which returns some JSX... then it's a component :)

**Review by sbegaudeau — CHANGES_REQUESTED** [review]
The build of the pull request is red and please try not to create pull requests between your own branches but instead against master so that we have a better idea of the real impact on the work of this branch.

File: backend/sirius-components-diagrams-layout/src/main/java/org/eclipse/sirius/components/diagrams/layout/ELKLayoutedDiagramProvider.java (hunk 1 above)
Comment (lfasani) [reply backend/sirius-components-diagrams-layout/src/main/java/org/eclipse/sirius/components/diagrams/layout/ELKLayoutedDiagramProvider.java:84]: I did not catch your first sentence. 

> What was the issue with the previous algorithm?

The frontend needs to display the label area centered on the node and each line part centered inside the node
I used "text-anchor:middle" to do that. But to make it work the label position needs to be positionned on the center of the node. I tried to apply `text-anchor:middle` only on tspan instead of the parent text but `text-anchor:middle` make still the line positionned on the label x position

File: backend/sirius-components-diagrams-layout/src/test/java/org/eclipse/sirius/components/diagrams/layout/incremental/DiagramLayoutTest.java (hunk 3 above)
Comment (lfasani) [reply backend/sirius-components-diagrams-layout/src/test/java/org/eclipse/sirius/components/diagrams/layout/incremental/DiagramLayoutTest.java:95]: Yes I need a real ELKDiagramConverter to have an non null ELKConvertedDiagram in the full layout process
But ok i will move this test to a new one out of incremental package

File: backend/sirius-components-diagrams/src/main/java/org/eclipse/sirius/components/diagrams/TextBoundsProvider.java (hunk 6 above)
Comment (lfasani) [reply backend/sirius-components-diagrams/src/main/java/org/eclipse/sirius/components/diagrams/TextBoundsProvider.java]: done

File: backend/sirius-components-emf/src/main/java/org/eclipse/sirius/components/emf/compatibility/properties/EStringIfDescriptionProvider.java (hunk 8 above)
Comment (lfasani) [reply backend/sirius-components-emf/src/main/java/org/eclipse/sirius/components/emf/compatibility/properties/EStringIfDescriptionProvider.java:42]: Ths is needed because the frontend component TextfieldPropertySection checks the typeName if it equals 'Textarea'

File: frontend/src/diagram/sprotty/DependencyInjection.ts (hunk 9 above)
Comment (lfasani) [reply frontend/src/diagram/sprotty/DependencyInjection.ts:73]: move to EditLabelUIWithInitialContent  class

File: frontend/src/diagram/sprotty/views/LabelView.tsx (hunk 11 above)
Comment (lfasani) [reply frontend/src/diagram/sprotty/views/LabelView.tsx:71]: I am not sure to understand what you expect.
I changed the code a little.
But perhaps you prefer something like this ?
```
{lines.length ===1? : text : lines.map()....}
```

File: backend/sirius-components-diagrams-layout/src/test/java/org/eclipse/sirius/components/diagrams/layout/incremental/DiagramLayoutTest.java (hunk 2 above)
Comment (lfasani) [reply backend/sirius-components-diagrams-layout/src/test/java/org/eclipse/sirius/components/diagrams/layout/incremental/DiagramLayoutTest.java:181]: done

File: frontend/src/diagram/sprotty/views/LabelView.tsx [hunk 12]
```diff
@@ -48,12 +49,40 @@ export class LabelView extends SLabelView {
...
+            }
+            return (
+              <tspan x="0" dy={fontSize}>
+                {line}
+              </tspan>
+            );
+          }
+        });
+      }
+    };
```
Comment (sbegaudeau) [inline frontend/src/diagram/sprotty/views/LabelView.tsx]: This component can be extracted and given props. Overwise its lifecycle is a bit odd since it is based on a closure

**Review by sbegaudeau — APPROVED** [review]
I'll just wait for the build and merge it after that
